core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java

```diff
@@ -76,14 +81,34 @@ public Boolean evaluate(ProjectConfig config, Map<String, ?> attributes) {
         Object userAttributeValue = attributes.get(name);
 
         if (!"custom_attribute".equals(type)) {
-            MatchType.logger.error(String.format("condition type not equal to `custom_attribute` %s", type));
+            logger.error("Audience condition \"{}\" has an unknown condition type: {}", this.toString(), type);
             return null; // unknown type
         }
         // check user attribute value is equal
         try {
-            return MatchType.getMatchType(match, value).getMatcher().eval(userAttributeValue);
+            Match matchType = MatchType.getMatchType(match, value).getMatcher();
+            Boolean result = matchType.eval(userAttributeValue);
+
+            if (!(matchType instanceof NullMatch) && result == null) {
+                if (!attributes.containsKey(name)) {
+                    //Missing attribute value
+                    logger.debug("Audience condition \"{}\" evaluated as UNKNOWN because no user attribute for \"{}\" key is passed", this.toString(), name);
+                } else {
+                    //if attribute value is not valid
+                    logger.warn(
+                        "Audience condition \"{}\" evaluated as UNKNOWN because the value for user attribute \"{}\" is inapplicable: \"{}\"",
+                        this.toString(),
+                        name,
+                        userAttributeValue);
+                }
+            } else if (matchType instanceof NullMatch) {
+                logger.error("Audience condition \"{}\" " + ((NullMatch) matchType).getMatchTypeError().toString(),
+                    this.toString(),
+                    match == null ? "legacy_custom_attribute" : match);
+            }
+            return result;
         } catch (NullPointerException np) {
-            MatchType.logger.error(String.format("attribute or value null for match %s", match != null ? match : "legacy condition"), np);
+            logger.error(String.format("attribute or value null for match %s", match != null ? match : "legacy condition"), np);
```

Comment:
Nit: This is a somewhat cryptic log message. @loganlinn @thomaszurkan-optimizely maybe we can audit our code to include all necessary null checks and then get rid of this `NullPointerException` catcher? Tho maybe it's not worthwhile if we don't expect to get here in practice...